### DISCUSSION TITLE
Possible to improve x-for performance with large lists?

### DISCUSSION THREAD

**Discussion by ragulka** [alpinejs/alpine#570] [4. General]
I ran into some performance issues with `x-for` while trying to implement a country dropdown in Alpine with keyboard support. The issue becomes apparent when trying to navigate the list with keyboard up/down arrows while having browser devtools open.

I've created a minimal test case for Alpine, Vue and Vanilla JS:
* https://codepen.io/ragulka/pen/QWyNZxv - Alpine
* https://codepen.io/ragulka/pen/GRoZYbZ - Vue
* https://codepen.io/ragulka/pen/QWyNJyQ - Vanilla JS

The Vue and Vanilla JS test cases doesn't seem to suffer from any kind of performance issues (even with devtools open), while Alpine becomes rather sluggish.

My guess is that this is partly due to Alpine not using Virtual DOM. However, given that even vanilla JS does not suffer from such issues, could any optimizations be made to how `x-for` is re-rendered?

**Comment by mblarsen** [comment]
Yeah, it is quite noticeable. Nice job on providing the codepens.

**Comment by ryangjchandler** [comment]
Hi @ragulka - the performance difference is quite noticeable. Great work on providing three different CodePens.

A couple of thing to note:

* Vue uses a Virtual DOM and will only update the necessary elements. This is something that Alpine doesn't currently do, since Alpine re-renders an entire component when any piece of data changes.

* Whenever a piece of data changes with Alpine, it needs to re-evaluate each expression and update the DOM accordingly, which will also have a larger performance impact compared to Vue and Vanilla JS. This isn't noticeable when the DevTools are closed, but when they're open, the DevTools is also evaluating and keeping track of each point in the stack. Alpine will have many more points to track compared to the Vanilla JS example.

* `x-for` has some overhead compared to a simple `Array.map`. When Alpine renders a list, it first checks if the array item being evaluated exists as an element, using the `:key` binding. If not, it then creates it. This allows for smart re-ordering of elements, without destroying each one. It also means that any elements that are not affected by the data change maintain their internal state (event listeners, attribute values, actual element properties). This overhead is worth it in my opinion, it makes the state of a component much more reliable and predictable.

**Reply by ragulka** [reply]
@ryangjchandler just to clarify - when you say that `Alpine re-renders an entire component when any piece of data changes` - does that mean if I have nested components, and a single property on the parent component changes (for example, a `show` boolean, to toggle a class on the component), it re-renders the parent and all the nested components, instead of just updating the class?

**Reply by dfallman** [reply]
Did this get answered?

**Reply by ekwoka** [reply]
@dfallman this post is from even before Alpine v3 which included basically complete rewrites to all these systems.

If you're seeing issues that seems related, I would advise making a new discussion and maybe referencing this but providing details on your issue.

Broadly, Alpine can handle quite large lists (few hundred at least) before noticeable issues should arise.

**Reply by dfallman** [reply]
Thanks, yes. We've had some similar issues to the above in some situations when having nestled x-for loops, but can create a separate discussion

**Reply by shashank-brightness** [reply]
+1 for this. Getting similar issue. Alpine x-for loop just stucks for 5 second before showing a list of 187 items.

**Reply by ekwoka** [reply]
5 seconds for 187 is much much too many.

I've used it in production for thousands and it isn't that bad.

**Reply by shashank-brightness** [reply]
@ekwoka I have set :key to javascript timestamp to ensure each time unique key is generated for all of them. Could that be a problem?

**Reply by shashank-brightness** [reply]
@ekwoka I have set fetch method to fetch data from api and it is taking first 4 seconds and then after I print `Date.now()` when  promise is resolved in console, alpine js operation starts from that point. I measured difference of that timestamp to the point where everything is rendered in the dom and it turned out to be exactly near to 12030 ms now.

**Reply by ekwoka** [reply]
YES!! that's a huge problem!

Specifically because they don't be unique timestamps!!! it's possible multiple are handled in one ms (and likely)

It will cause massive problems!!!

The other thing that can affect it is actually how much work needs to be done in each template. an x-for with a simple single element thing: easy. A massive DOM tree: hard

**Reply by shashank-brightness** [reply]
@ekwoka 
Thanks for suggestion.
I have added index to each timestamp `obj.key = Date.now() + index` so I am 100% sure about uniqueness but still I am getting that much delay in `x-for` even if I turned off other operations in JS. I forgot to mention that I had 290 elements and not 187 now. On first load it loads after that time. Subsequent operations are easier 

I have one boolean property named `isFeatured` along with each object and i am bifurcating them within further two more templates inside one whole template. Could this be a problem?

**Reply by shashank-brightness** [reply]
@ekwoka Yes I have massive DOM tree as you mentioned in last.  But what can we do in that case? I anyhow need to loop through with all operations and elements within `x-for` only for all array of shops. What do you recommend?

**Reply by ekwoka** [reply]
If you don't have actual ids, just don't put the key at all, and it'll use index.

Beyond that, I'd need to see more code to have any sense. Would be helpful if you made a separate discussion for your case.

**Reply by dfallman** [reply]
Would be interesting to understand what people have tried to do to speed up `x-for` especially when working with large lists. 

We have 1000s of lines and are experiencing a slowdown, but not to the point of it being useless. Each of our lines are also quite complex and need to circle in and out of multiple javascript functions to be rendered.

A couple of things we've found is that:

1) Nested `x-for` loops and `x-for` loops with `x-if`s tend to slow things down substantially. I assume that has to do with the lack of shadow DOM and something like Vue's computed properties. Both of these would be amazing to have with Alpine as well, even in the form of a choice / plugin. Is there any work undergoing/planned to add something like this to Alpine?

2) We found that use of Tailwind's `line-clamp-x` seems to slow the browser down quite dramatically. We're actually not sure why this is the case but we noticed a pretty dramatic increase in performance by just using something like `overflow-hidden` instead. This might be more of a Tailwind and/or browser issue, but could impact performance.

**Reply by ekwoka** [reply]
@dfallman I'm curious what you mean about vue's computed properties as it applies to this.

On 2. yes, line-clamp does some layout thrashing, so adding large amounts of elements with it can be an issue, and the way Alpine adds elements could probably be improved to handle this (swap all elements at once, instead of one at a time). Since the clamp has behaviors that require it to check the layout before being able to finalize what should render, and it would likely be doing it for every element added. Could be, maybe not.

There are some approaches I've been investigating to give simple apis for handling high numbers of components from a for loop trying to depend on just a few values that often don't actually change the result of the expression, like some kind of view into the data that reduces how often expressions need to be reevaluated, but haven't quite found something that seems elegant enough to put into Alpine core.

**Reply by dfallman** [reply]
@ekwoka that is perhaps specific to our implementation, where we have a master list of >1000s of objects that each needs to check with/be double-bound with ~10-ish js functions, sorted into a handful of different views, and then handed over to `x-for` to be displayed. Computed properties would likely speed up the first part of this rendering significantly. The second part of the rendering would likely be sped up by a virtual DOM.

I'm curious as to if/when a (perhaps optional?) virtual DOM is on the horizon for Alpine, or whether it's even been discussed? Listening to a few of @calebporzio podcast appearances you get a sense it's not going to happen (?) I think that's a shame in a way as it appears to hold back Alpine's usefulness for larger, more data-intensive projects.

We're basically looking to do a re-write in Vue/Nuxt.

**Reply by ekwoka** [reply]
Virtual Dom is almost definitely not on the table (though I don't speak for Caleb).

Virtual Dom just isn't actually faster. For example: Solidjs doesn't use a virtual Dom, and it's the fastest framework. Virtual Dom is pure overhead.

The things that can cause large loops to be slow wouldn't be changed with virtual Dom at all.

It's in how expressions are bound.


And yes, Alpine is just not meant for large data intensive projects, though it can be made to be pretty okay at it.

**Comment by andymerskin** [comment]
I wonder if a plugin could be written along the lines of [Tanstack Virtual](https://tanstack.com/virtual/v3), to handle this?

**Reply by shashank-brightness** [reply]
Any thoughts on this? as I am now even facing harder scrolling issue for alpine js array of objects in UI even when there is no nested loop inside. I did not know about 60 FPS concept earlier but it looks like it is same thing which prevents us to scroll naturally in chrome browser.

**Reply by ekwoka** [reply]
Thanks for this.

I might try whipping up an alpha plugin with that library and see how well it works for the case.

there may be factors that are a bit more unique to Alpine's use cases that would make a generalized virtualizer not good enough.

**Comment by calebporzio** [comment]
We've recently merged a PR to improve keyboard navigation performance in large lists drastically for the X-listbox and X-combobox components. Just a heads up.
